src/main/drivers/vtx_smartaudio.c

```diff
@@ -0,0 +1,780 @@
+                // XXX Have to implement chained request...
+            } else {
+                dprintf(("smartAudioSetPowerByIndex: V2 value %d\r\n", 
+                    saPowerTableV2[index].value));
+                buf[4] = saPowerTableV2[index].value;
+                buf[5] = CRC8(buf, 5);
+                saQueueCmd(buf, 6);
+            }
+        } else {
+            // Pit mode requested.
+            // Not implemented yet.
+        }
+    }
+}
+
+#ifdef SMARTAUDIO_EXTENDED_API
+void smartAudioSetMode(int mode)
+{
+    static uint8_t buf[6] = { 0xAA, 0x55, SACMD(SA_CMD_SET_MODE), 1 };
+
+    buf[4] = mode & 0x0f;
+    buf[5] = CRC8(buf, 5);
+
+    saQueueCmd(buf, 6);
+}
+#endif
+
+void smartAudioInit(void)
+{
+    portOptions_t portOptions;
+
+#ifdef SMARTPORT_DPRINF
+    // Setup debugSerialPort
+
+    debugSerialPort = openSerialPort(SERIAL_PORT_USART3, FUNCTION_NONE, NULL, 115200, MODE_RXTX, 0);
+    if (!debugSerialPort)
+        return;
+    setPrintfSerialPort(debugSerialPort);
+    dprintf(("smartAudioInit: OK\r\n"));
+#endif
+
+    portOptions = SERIAL_BIDIR|SERIAL_BIDIR_PP;
+
+    // XXX Fixed at UART2 fow the first cut
+    smartAudioSerialPort = openSerialPort(SERIAL_PORT_USART2, FUNCTION_NONE, NULL, 4800, MODE_RXTX, portOptions);
+
+    if (!smartAudioSerialPort) {
+        return;
+    }
+}
+
+#ifdef SMARTAUDIO_EXTENDED_API
+bool smartAudioIsReady(void)
+{
+    return (sa_vers != 0);
+}
+
+int smartAudioGetPowerTable(int *pTableSize, vtxPowerTable_t **pTable)
+{
+    switch (sa_vers) {
+    case 0:
+        return -1;
+
+    case 1:
+        if (pTableSize)
+            *pTableSize = 4;
+
+        if (pTable)
+            *pTable = saPowerTableV1;
+
+        break;
+
+    case 2:
+        if (pTableSize)
+            *pTableSize = 5;
+
+        if (pTable)
+            *pTable = saPowerTableV2;
+    }
+
+    return 0;
+}
+#endif
+
+#ifdef SMARTAUDIO_PITMODE_DEBUG
+void smartAudioPitMode(void)
+{
+    static int turn = 0;
+
+    if ((turn++ % 2) == 0) {
+        smartAudioSetMode(SA_MODE_SET_UNLOCK|SA_MODE_SET_PITMODE|SA_MODE_SET_IN_RANGE_PITMODE);
+    } else {
+        smartAudioSetMode(SA_MODE_SET_UNLOCK|SA_MODE_CLR_PITMODE);
+    }
+}
+#endif
+
+
+#define SCHED_PERIOD_DISARMED  (200*1000) // 200msec
+#define SCHED_PERIOD_ARMED (1000*1000) // 1sec, really want to make it none.
+
+void smartAudioProcess()
+{
+    static bool previousArmedState = false;
+
+    if (smartAudioSerialPort == NULL) {
+        return;
+    }
+
+    // Try reducing scheduling frequency (overkill?),
+    // and conserve as much cycles as we can.
+
+    bool armedState = ARMING_FLAG(ARMED) ? true : false;
+    bool armedStateChanged = armedState != previousArmedState;
+    previousArmedState = armedState;
+
+    if (armedStateChanged) {
+        if (armedState) {
+            // Reduce scheduling frequency
+
+            rescheduleTask(TASK_VTX_CONTROL, SCHED_PERIOD_ARMED);
```

Comment:
This change in scheduling frequency should be done by the caller. Driver should have no knowledge of scheduler.